agent/lm_agent/backend_utils/utils.py

```diff
@@ -0,0 +1,357 @@
+
+    logger.debug(f"Successfully loaded token from cache file {token_path}.")
+    return token
+
+
+def _write_token_to_cache(token: str):
+    """
+    Writes the token to the cache.
+    """
+    cache_dir = settings.CACHE_DIR
+    if not cache_dir.exists():
+        logger.warning(f"Cache directory does not exist {cache_dir}. Token won't be saved.")
+        return
+
+    token_path = settings.CACHE_DIR / TOKEN_FILE_NAME
+    try:
+        token_path.write_text(token)
+        token_path.chmod(0o600)
+        logger.debug(f"Successfully saved token to {token_path}")
+    except Exception as err:
+        logger.warning(f"Couldn't save token to {token_path}: {err}")
+
+
+def acquire_token() -> str:
+    """
+    Retrieves a token from OIDC based on the app settings.
+    """
+    logger.debug("Attempting to use cached token")
+    token = _load_token_from_cache()
+
+    if token is None:
+        logger.debug("Attempting to acquire token from OIDC")
+        oidc_body = dict(
+            audience=settings.OIDC_AUDIENCE,
+            client_id=settings.OIDC_CLIENT_ID,
+            client_secret=settings.OIDC_CLIENT_SECRET,
+            grant_type="client_credentials",
+        )
+        oidc_url = f"https://{settings.OIDC_DOMAIN}/protocol/openid-connect/token"
+        logger.debug(f"Posting OIDC request to {oidc_url}")
+        response = httpx.post(oidc_url, data=oidc_body)
+        LicenseManagerAuthTokenError.require_condition(
+            response.status_code == 200, f"Failed to get auth token from OIDC: {response.text}"
+        )
+        with LicenseManagerAuthTokenError.handle_errors("Malformed response payload from OIDC"):
+            token = response.json()["access_token"]
+
+        logger.debug("Successfully acquired auth token from OIDC")
+        _write_token_to_cache(token)
+
+    return token
+
+
+class AsyncBackendClient(httpx.AsyncClient):
+    """
+    Extends the httpx.AsyncClient class with automatic token acquisition for requests.
+    The token is acquired lazily on the first httpx request issued.
+
+    This client should be used for most agent actions.
+    """
+
+    _token: Optional[str]
+
+    def __init__(self):
+        self._token = None
+        super().__init__(base_url=settings.BACKEND_BASE_URL, auth=self._inject_token)
+
+    def _inject_token(self, request: httpx.Request) -> httpx.Request:
+        if self._token is None:
+            self._token = acquire_token()
+        request.headers["authorization"] = f"Bearer {self._token}"
+        return request
+
+
+async def check_backend_health():
+    """
+    Hit the API's health-check endpoint to make sure the API is available.
+    """
+    async with AsyncBackendClient() as backend_client:
+        resp = await backend_client.get("/lm/health")
+    if resp.status_code != 204:
+        logger.error("license-manager-backend health-check failed.")
+        raise LicenseManagerBackendConnectionError("Could not connect to the backend health-check endpoint")
```

Comment:
so, because our deployments are canary deployments, this would actually never happen. However, to make it realistic for people that use other type of deployments, I recommend setting up some retries here, as well as some timeout. This function can raise unrealistic behaviours if these parameters I mentioned cannot be set along side the load balancing parameters

Reply:
you can disregard my comment since the cases I'm thinking are not relevant now